nselib/smb2.lua

```diff
@@ -0,0 +1,410 @@
+  SMB2_GLOBAL_CAP_LEASING             = 0x00000002,
+  SMB2_GLOBAL_CAP_LARGE_MTU           = 0x00000004,
+  SMB2_GLOBAL_CAP_MULTI_CHANNEL       = 0x00000008,
+  SMB2_GLOBAL_CAP_PERSISTENT_HANDLES  = 0x00000010,
+  SMB2_GLOBAL_CAP_DIRECTORY_LEASING   = 0x00000020,
+  SMB2_GLOBAL_CAP_ENCRYPTION          = 0x00000040,
+  -- Context Types
+  SMB2_ENCRYPTION_CAPABILITIES        = 0x0002,
+  SMB2_PREAUTH_INTEGRITY_CAPABILITIES = 0x0001
+}
+
+for i, v in pairs(command_codes) do
+  command_names[v] = i
+end
+for i, v in pairs(smb2_values) do
+  smb2_values_codes[v] = i
+end
+
+---
+-- Creates a SMB2 SYNC header packet.
+--
+-- SMB2 Packet Header - SYNC: 
+-- * https://msdn.microsoft.com/en-us/library/cc246529.aspx
+--
+-- @param smb The SMB object associated with the connection.
+-- @param command The SMB2 command to execute.
+-- @param overrides Overrides table.
+-- @return header The encoded SMB2 SYNC header.
+---
+function smb2_encode_header_sync(smb, command, overrides)
+  overrides = overrides or {}
+
+  local sig = "\xFESMB" -- SMB2 packet 
+  local structureSize = 64 -- SYNC header structure size
+  local flags = 0 -- TODO: Set flags that will work for all dialects
+
+  -- Increase the message id
+  if smb['MessageId'] then
+    smb['MessageId'] = smb['MessageId'] + 1
+  end
+
+  -- Header structure
+  local header = string.pack("<c4 I2 I2 I4 I2 I2 I4 I4 I8 I4 I4 I8 c16",
+    sig,                                -- 4 bytes: ProtocolId
+    structureSize,                      -- 2 bytes: StructureSize. Must be 64.
+    (overrides['CreditCharge'] or 0),   -- 2 bytes: CreditCharge. 
+    (overrides['Status'] or 0),         -- 4 bytes: (ChannelSequence/Reserved)/Status. 
+    command,                            -- 2 bytes: Command.
+    (overrides['CreditR'] or 0),        -- 2 bytes: CreditRequest/CreditResponse.
+    (overrides['Flags'] or flags),      -- 4 bytes: Flags. TODO
+    (overrides['NextCommand'] or 0),    -- 4 bytes: NextCommand.
+    (overrides['MessageId'] or smb['MessageId'] or 0),  -- 8 bytes: MessageId.
+    (overrides['Reserved'] or 0),              -- 4 bytes: Reserved.
+    (overrides['TreeId'] or smb['TreeId'] or 0),        -- 4 bytes: TreeId.
+    (overrides['SessionId'] or smb['SessionId'] or 0),  -- 8 bytes: SessionId.
+    (overrides['Signature'] or '1234567890123456')     -- 16 bytes: Signature.
+    )
+
+  return header
+end
+
+---
+-- Sends a SMB2 packet 
+-- @param smb        The SMB object associated with the connection
+-- @param header     The header encoded with <code>smb_encode_sync_header</code>.
+-- @param data       The data.
+-- @param overrides  Overrides table.
+-- @return (result, err) If result is false, err is the error message. Otherwise, err is
+--        undefined
+---
+function smb2_send(smb, header, data, overrides)
+  overrides = overrides or {}
+  local body               = header .. data
+  local attempts           = 5
+  local status, err
+
+  local out = string.pack(">I<c" .. #body, #body, body)
```

Comment:
Is this really big-endian? I thought all SMB stuff was little-endian. Also, you need to specify a length in bytes, otherwise it uses "native" length which may vary on some platforms. Lastly, use the "S" family of [format strings](https://www.lua.org/manual/5.3/manual.html#6.4.2) for length-prefixed strings. I'd guess this is supposed to be `string.pack("<S4", body)`, but you should verify.